activestorage/test/models/attached/many_test.rb

```diff
@@ -667,6 +667,19 @@ def highlights
     end
   end
 
+  test "attaches two files using association setter and #attach" do
+    append_on_assign do
+      user = User.new(name: "Tina", highlights: [fixture_file_upload("image.gif")])
+      user.highlights.attach fixture_file_upload("racecar.jpg")
+      user.save
+
+      user.reload
+      assert_equal 2, user.highlights.count
+      assert_equal "image.gif", user.highlights.first.filename.to_s
+      assert_equal "racecar.jpg", user.highlights.second.filename.to_s
+    end
+  end
+
```

Comment:
The following test

```ruby
  test "attaches two files using association setter and #attach" do
    append_on_assign do
      user = User.new(name: "Tina", highlights: [fixture_file_upload("image.gif")])
      user.highlights = [fixture_file_upload("racecar.jpg")]
      user.save

      user.reload
      assert_equal 2, user.highlights.count
      assert_equal "image.gif", user.highlights.first.filename.to_s
      assert_equal "racecar.jpg", user.highlights.second.filename.to_s
    end
  end
```

would pass before this PR and fail with this PR. It's not respecting the `append on assign` behavior while the record is dirty, thoughts?

Reply:
You are right. This test cases does fail. I would expect it to attach both blobs. So far what I found was since we are using `association_name=` setter methods which directly takes attachables as argument. It replaces with new attachables passed as argument whereas when used with `attach` sends attachable argument considering `(change&.attachables || blobs) + attachables.flatten`

Will dig deeper but somewhere I feel like if you use `association_name=` when record is dirty it shouldn't append. WDYT?

Reply:
Yeah, I don't have a strong opinion.. It's hard to tell how it should work since the setter method appending attachments is weird by itself 😅 

I guess that if we can keep existing behavior it'd be better since it'd be easier for people to upgrade, if it's not possible then I guess I'd be fine introducing this change in behavior

Reply:
@abhaynikam the behaviour with `replace_on_assign_to_many = false` is intentionally for backwards compatibility (see #36716). Give this PR changes that behaviour I don't think we can merge it as is and I'd rather see that setting deprecated in Rails 7.0 because as you can see it produces confusing behaviour (that's also George's intention from the original PR).

Reply:
@pixeltrix I have updated the PR without breaking the backward compatibility. Could you please have another look at the PR? Tested it locally on master branch as well. cc/ @zzak @santib

Reply:
@abhaynikam does this restore the behaviour to what it was in 5.2 or is it different? If it's fixing a regression from 5.2 then it's okay. Might be easiest to create the same test cases using two standalone examples - one using 5.2 and one using your changes.

Reply:
@pixeltrix I have added two specs covering both scenarios. IMO, it does restore the old behaviour letting user append the attachment to the collection when `replace_on_assign_to_many = false` config is disabled. Ref: https://github.com/rails/rails/pull/42455/files#diff-e346e335288f81186850ba1a547cc88e1009ac51b5194769dfa1cc615aaff75fR777

Also, this resolve the issue regarding duplicating the attachment blobs when Active Storage setter methods(i.e: `association_name=` and `attach`) are used in a combination before transaction is committed.

Sorry if I am failing to understand the comment 😊

Reply:
What I'm saying is that de-duplicating things may actually be breaking backwards compatibility. An app may be assuming that behaviour and doing its own hack to fix it and we may break their fix. That's not necessarily a reason not to fix it but we should know what behaviour we're changing so that we're fully informed.

I'm suggesting we create two examples using the [bug report][1] template, one with 5.2 and one with your branch so that we can compare the differences.

[1]: https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_storage_gem.rb